4단계 - 블랙잭(베팅)

giwankim opened the pull request:
결국 블랙잭 미션 마지막 단계까지 왔습니다 🔥 

미리 생각해 두었던 기능이라 조금 용이하게 구현한 것 같습니다. 💪 

끝까지 잘 부탁드리겠습니다. 🙇

giwankim committed:
refactor: 불필요한 enum class 를 제거했다.

참조: https://github.com/next-step/kotlin-blackjack/pull/781#discussion_r1875445307

giwankim committed:
refactor: 초기화 카드 숫자 상수를 `Hand` 클래스로 옮겼다.

참조: https://github.com/next-step/kotlin-blackjack/pull/781#discussion_r1875438513

giwankim committed:
refactor: 딜러와의 승부 결과 계산을 `PlayerOutcome` enum class 에 위임한다.

참조: https://github.com/next-step/kotlin-blackjack/pull/781#discussion_r1875436229

giwankim committed:
feat: 플레이어의 베팅 금액을 입력받는다.

giwankim committed:
feat: 플레이어는 베팅을 걸 수 있다.

giwankim committed:
fix: 줄바꾸면서 베팅 금액을 입력받는다.

giwankim committed:
feat: 블랙잭을 제외한 최종 수익을 계산한다.

giwankim committed:
feat: 딜러와 플레이어들의 최종 수익을 출력한다.

giwankim commented:
~~앗, 기능을 하나 빠뜨렸군요 😱~~
~~바로 수정하고 이 메시지 지우겠습니다.~~

giwankim committed:
feat: 플레이어가 블랙잭이고 딜러가 블랙잭이 아닌 경우 베팅 금액의 1.5배를 딜러에게서 받는다.

wisemuji commented on README.md:
블랙잭 규칙 중 다음 규칙도 적용하시고 테스트 코드를 추가해보시면 어떨까요?

> 플레이어의 턴에서 21이 되면(블랙잭) 더 이상 해당 플레이어에게 카드를 받지 않고 다음 턴으로 넘어간다.

giwankim replied:
ㅋㅋ 일반적으로 카지노는 이렇게 나이스하지 않은 것 같은데 연습 삼아 구현해보겠습니다. 

두 장의 카드를 받아서 21점인 경우만 (AK, AQ, AJ, AT) 블랙잭으로 자동 종료되고 카드를 뽑아서 21점이 되는 경우 계속 힛하면 버스트하는 것으로 알고 있습니다.

wisemuji commented on src/main/kotlin/blackjack/domain/PlayerOutcome.kt:
```diff
@@ -1,7 +1,33 @@
+enum class PlayerOutcome(
+    private val ratio: BigDecimal,
+) {
+    BLACKJACK(BigDecimal(1.5)),
```
적절한 객체에게 배팅 금액 계산 비율을 들고 있을 수 있도록 위임해주시고, 이를 판별할 수 있는 함수도 잘 만들어주셨네요 👍 

1.5배 요구 사항도 잘 만족하도록 구현해주셨네요 👍

wisemuji commented on src/main/kotlin/blackjack/application/CreateGameCommand.kt:
```diff
@@ -0,0 +1,16 @@
+) {
+    init {
+        require(names.size == bets.size) {
+            "이름들과 베팅들의 숫자가 일치해야 합니다."
```
중복 제거를 위해 이름은 Set 자료형을 활용하는 것도 고려해볼 수 있을 것 같아요!

wisemuji commented on src/main/kotlin/blackjack/application/BlackjackService.kt:
```diff
@@ -1,16 +1,17 @@
-        val players = Players.from(names)
-        val game = Game(players, deck)
+    fun createGame(command: CreateGameCommand): Game {
+        val bets = command.bets.map { Bet(it) }
```
사소하지만 아래와 같이 작성해볼 수도 있겠네요!
```suggestion
        val bets = command.bets.map(::Bet)
```

wisemuji commented on src/main/kotlin/blackjack/domain/Hand.kt:
```diff
@@ -37,5 +37,6 @@ open class Hand(
         private const val ACE_ELEVEN_THRESHOLD = 11
         private const val ACE_EXTRA_VALUE = 10
         private const val BLACKJACK_VALUE = 21
+        const val INITIAL_HAND_SIZE = 2
```
👍  도메인에서 들고 있도록 옮겨주셨네요

wisemuji commented on src/main/kotlin/blackjack/domain/Players.kt:
```diff
@@ -43,9 +43,18 @@ class Players(
-    fun result(dealer: Dealer): List<PlayerResult> {
+    fun placeBets(bets: List<Bet>) {
+        requireBetsSizeEqualsRosterSize(bets)
+        roster.zip(bets).forEach { (player, bet) -> player.placeBet(bet) }
```
https://kotlinlang.org/docs/coding-conventions.html#wrap-chained-calls
코틀린 공식 문서에 따르면 함수 체인의 경우 아래와 같이 체인별 개행을 해주어야 합니다.
(일반적으로 체인의 첫 번째 호출에는 개행이 있어야 하지만 개행하지 않는 편이 더 합리적이라면 생략해도 괜찮습니다)

```suggestion
        roster.zip(bets)
            .forEach { (player, bet) -> player.placeBet(bet) }
```

wisemuji commented on src/test/kotlin/blackjack/domain/PlayerOutcomeTest.kt:
```diff
@@ -0,0 +1,134 @@
+        profit shouldBe BigDecimal.ZERO
+    }
+
+    @Test
+    fun `블랙잭인 경우 1_5 배의 수익을 본다`() {
+        val bet = Bet(10_000L)
+        val outcome = PlayerOutcome.BLACKJACK
+
+        val profit = outcome.profit(bet)
+
+        profit shouldBe BigDecimal(15_000L)
+    }
```
테스트 시나리오까지 👍

wisemuji reviewed (approved):
안녕하세요 지완님!

4단계 역시 구현을 잘 해주셨습니다! 마지막까지 꼼꼼한 테스트 시나리오를 작성해주셔서 리뷰하기 편했습니다.
고민해보시면 좋을만한 코멘트를 조금 달았는데요, 큰 구조 변경은 없을 것 같아서 다음 리뷰 요청에는 PR을 머지할 수 있을 것 같습니다 🙂 

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요!

wisemuji reviewed (changes requested):
안녕하세요 지완님!

4단계 역시 구현을 잘 해주셨습니다! 마지막까지 꼼꼼한 테스트 시나리오를 작성해주셔서 리뷰하기 편했습니다.
고민해보시면 좋을만한 코멘트를 조금 달았는데요, 큰 구조 변경은 없을 것 같아서 다음 리뷰 요청에는 PR을 머지할 수 있을 것 같습니다 🙂

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요!

wisemuji commented:
실수로 Approve을 눌렀네요 😅 
Request changes가 맞습니다!

giwankim committed:
refactor: 함수 체인 시 체인별 개행을 해주었다.

[공식 문서 컨벤션](https://kotlinlang.org/docs/coding-conventions.html#wrap-chained-calls)에서도 추천하는 개행 방법이다.

참조: https://github.com/next-step/kotlin-blackjack/pull/792#discussion_r1879256116

giwankim committed:
refactor: callable (function) reference 를 사용해서 가독성을 높였다.

giwankim committed:
refactor: DTO 에서도 중복 제거 확인되도록 이름을 Set 자료형으로 입력받는다.

https://github.com/next-step/kotlin-blackjack/pull/792#discussion_r1879249940

giwankim committed:
feat: 플레이어 턴에서 합이 21이 되면, 더 이상 해당 플레이어에게 카드를 받지 않고 다음 턴으로 넘어간다.

참조: https://github.com/next-step/kotlin-blackjack/pull/792#discussion_r1879234957

giwankim committed:
refactor: 화면 출력 함수들을 확장 함수로 변경해서 가독성 개선했다.

giwankim committed:
refactor: getter 만 있는 property 는 한줄로 표기한다.

giwankim committed:
refactor: 테스트를 위한 보조 생성자 대신 fixture 메서드를 추가했다.

giwankim committed:
feat: 플레이어의 상태를 나타내는 인터페이스를 도출한다.

giwankim committed:
feat: 딜러로 부터 카드를 받는 중인 DEALING 상태 구현

giwankim committed:
feat: 행동(힛, 스탠드)을 할 수 있는 READY 상태 구현

giwankim committed:
feat: 종단 상태 (STAND, BUSTED, BLACKJACK) 구현

giwankim committed:
refactor: Finished 추상 클래스

giwankim committed:
refactor: Playing 추상 클래스를 추출했다.

giwankim committed:
refactor: `State` 인터페이스를 `Player` 가 사용하도록 리팩터링했다.

giwankim committed:
refactor: 딜러가 `State` 인터페이스를 사용하도록 리팩터링했다.

giwankim committed:
refactor: `Player` 클래스에서 베팅이 nullable 이 아니게 변경한다.

value class 는 lateinit 이 가능하지 않기에 `Bet` 을 data class 로 변경한다.

giwankim committed:
refactor: 수익 계산 리팩터링

wisemuji commented on src/main/kotlin/blackjack/domain/State.kt:
크 수업에서 소개된 상태 패턴을 잘 적용해주셨네요 👍 어려운 점은 없으셨나요?
IntelliJ로 구조도를 그려보니 아래와 같았어요.

<img width="801" alt="image" src="https://github.com/user-attachments/assets/461d3687-eddf-4c3b-89a5-288a0c21fab0" />

같은 블랙잭 요구사항이여도 상태 패턴을 어떻게 적용하느냐에 따라서 의존 관계나 구조도가 천차만별인데요,
테스트 시나리오를 꼼꼼히 작성해주셔서 이해하기 편했습니다 👍

wisemuji commented on src/main/kotlin/blackjack/domain/Stand.kt:
```diff
@@ -0,0 +1,16 @@
+    override val profitRatio: BigDecimal = BigDecimal.ONE
+
+    init {
+        require(hand.size >= Hand.INITIAL_HAND_SIZE) { "스탠드는 세 장 이상의 카드로 이루어져야 합니다." }
```
이건 저도 몰랐던 규칙이네요 😮

giwankim replied:
어 이건, 코드는 맞지만, 에러 국문 메시지가 잘못된 것 같습니다. 의도는 스탠드는 카드 두개를 발급 받은 다음부터 할 수 있는 행위라는 검증이었는데요....

wisemuji commented on src/main/kotlin/blackjack/domain/Blackjack.kt:
```diff
@@ -0,0 +1,14 @@
+class Blackjack(
+    hand: Hand,
+) : Finished(hand) {
+    override val profitRatio: BigDecimal = BigDecimal(1.5)
```
1.5배 요구사항을 이렇게 캡슐화하셨군요!

이제 불필요한 조건문을 제거하고, 다형성을 활용할 수 있게 되었습니다 🙂 

참고로 제가 DM방에 보내드렸던 [창시자 앨런 케이가 말하는, 객체 지향 프로그래밍의 본질](https://velog.io/@eddy_song/alan-kay-OOP)에서도 이야기하는 본질을 잘 이해하고 구현해주신 것 같아요

wisemuji commented on src/main/kotlin/blackjack/domain/Hand.kt:
```diff
@@ -3,25 +3,26 @@ package blackjack.domain
-    val cards = cards.toMutableList()
-
-    constructor(vararg cards: Card) : this(cards.toList())
+    val cards: MutableList<Card> = cards.toMutableList()
```
현재 구조에서는 외부에서 Hand의 cards를 직접 꺼내어 조작할 수 있는데요,
내부에 mutable 값을 캡슐화하도록 [backing property](https://kotlinlang.org/docs/properties.html) 적용을 고려해보셔도 좋을 것 같아요 🙂

wisemuji reviewed (approved):
안녕하세요 지완님! 마지막까지 계속 코드를 리팩터링하시는 과정을 이메일 알람으로 받았는데요, 리뷰 반영을 너무너무 잘 해주시고 상태 패턴도 잘 적용해주셨습니다 👍 

덕분에 저 또한 코드를 보며 좋은 인사이트를 얻어가는 시간이였습니다!
개발자로 성장하는 과정에 도움이 되었길 바라며 이번 미션은 마무리하겠습니다.
나머지 미션도 꼭 완주하시길 응원하겠습니다!
